third_party/gtp_ovs/ovs-gtp-patches/2.15/0023-QFI-Support-in-OVS.patch

```diff
@@ -0,0 +1,929 @@
+     case OFPACT_CHECK_PKT_LARGER:
+     case OFPACT_DELETE_FIELD:
++    case OFPACT_SET_TUNNEL_QFI:
+     default:
+         return false;
+     }
+@@ -9403,6 +9474,8 @@ ofpacts_parse__(char *str, const struct ofpact_parse_params *pp,
+             error = parse_pop_vlan(pp);
+         } else if (!strcasecmp(key, "set_tunnel64")) {
+             error = parse_set_tunnel(value, NXAST_RAW_SET_TUNNEL64, pp);
++        } else if (!strcasecmp(key, "qfi")) {
++            error = parse_set_tunnel_qfi(value, pp);
+         } else if (!strcasecmp(key, "load")) {
+             error = parse_reg_load(value, pp);
+         } else if (!strcasecmp(key, "bundle_load")) {
+diff --git a/lib/packets.h b/lib/packets.h
+index 6edf85f05..02405a60c 100644
+--- a/lib/packets.h
++++ b/lib/packets.h
+@@ -1427,7 +1427,8 @@ static inline ovs_be32 get_erspan_ts(enum erspan_ts_gra gra)
+ #define GTPU_MSGTYPE_GPDU   255 /* User Payload. */
+ 
+ enum {
+-    GTP_METADATA_V1
++    GTP_METADATA_V1,
++    GTP_METADATA_EXT_HDR_DATA
+ };
+ 
+ #define GTP_FLAGS_SEQ   0x02
+@@ -1436,8 +1437,11 @@ struct gtpu_metadata {
+     uint8_t ver;
+     uint8_t flags;
+     uint8_t msgtype;
++    uint8_t qfi;
++    uint8_t header_data_len;
++    uint8_t data[];
+ };
+-BUILD_ASSERT_DECL(sizeof(struct gtpu_metadata) == 3);
++BUILD_ASSERT_DECL(sizeof(struct gtpu_metadata) == 5);
+ 
+ /*
+  * GTP flags:
+@@ -1463,6 +1467,14 @@ struct gtpuhdr_opt {
+ };
+ BUILD_ASSERT_DECL(sizeof(struct gtpuhdr_opt) == 4);
+ 
++struct gtpu_ext_hdr {
++    uint8_t len;
++    uint8_t pdu_type;
++    uint8_t qfi;
++    uint8_t next_type;
++};
++BUILD_ASSERT_DECL(sizeof(struct gtpu_ext_hdr) == 4);
++
+ /* VXLAN protocol header */
+ struct vxlanhdr {
+     union {
+diff --git a/tests/system-layer3-tunnels.at b/tests/system-layer3-tunnels.at
+index 6e5e8a01a..5a5808dcf 100644
+--- a/tests/system-layer3-tunnels.at
++++ b/tests/system-layer3-tunnels.at
+@@ -1046,3 +1046,114 @@ dnl sleep 1000
+ OVS_TRAFFIC_VSWITCHD_STOP
+ AT_CLEANUP
+ 
++
++AT_SETUP([layer3 - Qfi value set for uplink over GTP])
++OVS_TRAFFIC_VSWITCHD_START([set Bridge br0 other-config:hwaddr="00:12:34:56:78:bb"])
++OVS_CHECK_GTP_L3()
++OVS_CHECK_MIN_KERNEL(4,10)
++
++ADD_BR([br-underlay])
++
++ADD_NAMESPACES(at_ns0)
++
++dnl Set up underlay link from host into the namespace using veth pair.
++ADD_VETH(p0, at_ns0, br-underlay, "172.31.1.1/24")
++AT_CHECK([ip addr add dev br-underlay "172.31.1.100/24"])
++AT_CHECK([ip link set dev br-underlay up])
++AT_CHECK([modprobe vport_gtp])
++
++dnl Set up tunnel endpoints on OVS outside the namespace and with a native
++dnl linux device inside the namespace.
++
++ADD_OVS_TUNNEL([gtpu], [br0], [at_gtp0], [172.31.1.1], [10.1.1.2/24])
++AT_CHECK([ip neigh add 10.1.1.1 lladdr 00:12:34:56:78:aa dev br0])
++NS_CHECK_EXEC([at_ns0], [gtp-link add at_gtp1 --sgsn &], [0], [ignore])
++dnl kernel 4.9
++NS_CHECK_EXEC([at_ns0], [gtp-tunnel add at_gtp1 v1 0 0 10.1.1.1 172.31.1.100], [0], [ignore], [ignore])
```

Comment:
There is no need for gtp-tunnel , when you use `gtp-packet.py` to send out packets.

Reply:
Yes, not required. Removed.